docs/source/tutorials/chat.rst

```diff
@@ -0,0 +1,284 @@
+
+When to use or not to use a prompt template is governed by what your desired inference
+behavior is. You should use a prompt template in the following scenarios:
+
+1. You are running inference on the base model and it was pre-trained with a prompt
+template
+2. You want to prime a fine-tuned model to expect a certain prompt structure on inference
+for a specific task
+
+It is not strictly necessary to fine-tune with a prompt template, but generally you
+want the model to perform some sort of task, which will require some formatting of
+the prompt.
+
+For example, the :class:`~torchtune.data.SummarizeTemplate` provides a lightweight
+structure to prime your fine-tuned model for prompts asking to summarize text.
+
+.. code-block:: python
+
+    f"Summarize this dialogue:\n{dialogue}\n---\nSummary:\n"
+
+You can fine-tune LLaMA2 with this template even though the model was originally pre-trained
+with the :class:`~torchtune.data.Llama2ChatFormat`, as long as this is what the model
+sees during inference. The model should be robust enough to adapt to a new template.
+
+Special tokens in LLaMA3
+------------------------
+
+LLaMA3 `overhauled <https://llama.meta.com/docs/model-cards-and-prompt-formats/meta-llama-3>`_
+the special tokens and prompt templating from LLaMA2 to better support multiturn conversations.
+
+.. code-block:: text
+
+    <|begin_of_text|><|start_header_id|>system<|end_header_id|>
+
+    {{ system_prompt }}<|eot_id|><|start_header_id|>user<|end_header_id|>
+
+    {{ user_message_1 }}<|eot_id|><|start_header_id|>assistant<|end_header_id|>
+
+    {{ model_answer_1 }}<|eot_id|><|start_header_id|>user<|end_header_id|>
+
+Now, all the additional tags are special tokens, and no prompt template is required.
+You can see how they get encoded as their own token IDs below:
+
+.. code-block:: python
+
+    from torchtune.modules.tokenizers import TikTokenTokenizer
+
+    tokenizer = TikTokenTokenizer("/tmp/Meta-Llama-3-8B/original/tokenizer.model")
+    print(tokenizer._encode_special_token("<|begin_of_text|>"))
+    # 128000
+    print(tokenizer._encode_special_token("<|eot_id|>"))
+    # 128009
+
+The best part is - all these special tokens are handled purely by the tokenizer.
+That means you won't have to worry about messing up any required prompt templates!
+
+Fine-tuning on a custom chat dataset
+------------------------------------
+
+Let's test our understanding by trying to fine-tune the LLaMA3-8B model with a custom
+chat dataset. We'll walk through how to set up our data so that it can be tokenized
+correctly and fed into our model.
+
+We'll be using the `LIMA <https://arxiv.org/pdf/2305.11206.pdf>`_ dataset, a small,
+carefully crafted dataset from Q&A selected for responses in the style of a helpful
+AI assistant. Looking at the :class:`~torchtune.datasets.ChatDataset` class, you'll
+notice several parameters you'll need to provide to configure it:
+
+.. code-block:: python
+
+    class ChatDataset(Dataset):
```

Comment:
Overall I like this section. Still a couple of comments: (1) I think it can be a bit overwhelming to just provide the general `ChatDataset` API from the get-go without a ton of context. Why not start with a single sample, show the format we want to get it into for tokenization, then walk through each of the necessary steps to get there? Then use that to motivate the exact definition of lima_dataset given below.

Reply:
That's a sound approach